Build names and links at runtime - no more sqlite

cpuguy83 opened the pull request:
This stores all names and link info in memory at runtime rather than relying on sqlite.

Please test thoroughly.
I added some tests, but this is a big change!

Fixes #17691

cpuguy83 commented:
phew, these rebases are rough.

cpuguy83 commented:
And realizing that hostconfig.Links was being set to nil before, so this information is only in sqlite...

LK4D4 commented:
−158,972

LK4D4 commented on hack/make/validate-lint:
```diff
@@ -76,6 +76,7 @@ packages=(
 	pkg/pubsub
 	pkg/random
 	pkg/reexec
+  pkg/registrar
```
something with formatting

LK4D4 commented:
Looks cool
windows failures might be related though

tiborvass commented:
Wow this is epic. Thanks so much @cpuguy83 !

Is there any undesirable side effect you can think of?

tianon commented:
Maybe an obvious question, but does this mean that links won't survive a daemon restart, or were they still stored on-disk somewhere else besides SQLite?

cpuguy83 commented:
Adding some context this from IRC convo:
Yes there is an issue here:
Currently `hostConfig.Links` is whipped out the first time it is processed.
So the only place links info is stored currently is in the sqlite db... so we can't just remove it here, it will require some migration code to insert links back into containers' hostconfig from sqlite.

cpuguy83 committed:
Build names and links at runtime

Don't rely on sqlite db for name registration and linking.
Instead register names and links when the daemon starts to an in-memory
store.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

swernli commented:
Confirmed that this works with Windows containers (both existing images after updating just docker.exe and on a fresh install).  Not sure why Windows Jenkins run is failing, but it works fine in manual tests.

LGTM

icecrime commented:
Ping @mavenugo @mrjana: can you please look into this in light of your plans regarding links deprecation (or not)?

thaJeztah commented:
ping @mavenugo @mrjana could you have a look at https://github.com/docker/docker/pull/16032#issuecomment-146645967 now that 1.9 has been release (and you hopefully have a bit more time :heart:)

calavera commented:
we cannot remove links from one version to the other. 

If we're really going to remove links we should deprecated them in 1.10 to completely remove sqlite3 and the code related in 1.12.

If we're going to keep links as they are, we can make the transition to in-memory links in 1.10, loading from sqlite to memory in that version, and remove the dependency in 1.11.

If we're going to keep links with a different implementation under the hood, we should make sure we remove sqlite from the dependencies.

cpuguy83 commented:
@calavera unfortunately we are stuck with sqlite in any case since `HostConfig.Links` is cleared out once read, currently. So the only place the link info is stored is in sqlite... so basically as long as we want to support upgrading from \<docker version with sqlite links\> --> \<docker version with something else\> it'll need to stay around.

thaJeztah commented:
@cpuguy83 should we close this for now or is this still an option? I see a discussion around removal of links, which is not yet clear when that'll happen

cpuguy83 commented:
@thaJeztah The overall goal is still valid, don't use sqlite for storing things on disk which we can build at runtime easily enough.
This should ultimately be much more efficient up to a certain point.

LK4D4 commented:
@cpuguy83 need rebase
Did you add migration logic?

cpuguy83 commented:
@LK4D4 Not yet.
Rebased for now, will work on migration.

cpuguy83 committed:
Add migration from sqlite links back to hostConfig

Before #16032, once links were setup
in the sqlite db, hostConfig.Links was cleared out.
This means that we need to migrate data back out of the sqlite db and
put it back into hostConfig.Links so that links specified on older
daemons can be used.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Added migration logic and test for migration.

cpuguy83 commented:
Errors look legit, investigating.

cpuguy83 commented:
Ok, all fixed.

tiborvass commented:
@cpuguy83 I can't tell if the in-memory state of links is restored on daemon boot. Can you confirm?
Otherwise, design +1 with migration.

/cc @mavenugo

cpuguy83 commented:
It is, and there are a couple of tests to ensure it.

tiborvass commented:
cool thanks, then this needs a rebase :)

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

cpuguy83 commented:
lol -- love it.

cpuguy83 commented:
omg, 3rd rebase today.

cpuguy83 commented:
Ok, this is ready.

tiborvass commented:
sorry @cpuguy83 :( needs rebase again, i want this

cpuguy83 commented:
Rebased

thaJeztah commented:
sorry @cpuguy83, looks like this needs another rebase :cry:

cpuguy83 commented:
Rebased

cpuguy83 commented:
Rebased and addressed comments.

calavera commented:
LGTM

calavera commented:
@cpuguy83 you might want to fix the message in your latest commit `s/hostConifg/hostConfig` :laughing:

cpuguy83 commented:
It's how I spell it quite frequently :)

cpuguy83 commented:
Fixed commit message.

mavenugo commented:
@cpuguy83 @icecrime @thaJeztah as discussed in the Alias proposal, we are not deprecating links. Infact the fix for retaining the links in hostConfig is much appreciated due to the nature of the decision of reusing links as the way to provide alias support for containers in user-defined networks.

I haven't spent time testing this PR manually yet. But the code LGTM.

Given the nature of changes and potential of the impact, it is better to do some additional migration tests (in addition to all the tests that is in Integration-cli).

@cpuguy83 can you add please some test-cases and results to make sure we are not regressing particularly upgrade scenarios.

cpuguy83 commented:
Rebased.

@mavenugo Do you have any suggestions for migration tests beyond what's in https://github.com/cpuguy83/docker/blob/remove_sqlite_dep/daemon/links_test.go?

thaJeztah commented:
darn, @cpuguy83 needs a rebase again :cry: this one is mentioned in the review-session e-mail so hopefully we can get this in asap to safe you from rebase hell

mavenugo commented:
@cpuguy83 sorry. missed your last comment. I was thinking of cases where user upgrades to 1.10, downgrades to 1.9 or earlier and again upgrades to 1.10. I know your change to retaining the hostConfig.Links will come to help us. But just to make sure we have a test will give us a peace of mind.

cpuguy83 commented:
A graceful downgrade is not something that I think is really possible here, and is also not something we've ever really guaranteed.

calavera commented:
I might be wrong but, the graceful downgrade is guaranteed unless we remove `linkgraph.db`. Obviously, if people add new containers to 1.10 and then downgrade those names will be lost. But old container names will be preserved in the database as long as the database file exists. That's guarantee enough for me.

cpuguy83 commented:
That's fine, I updated to make it not remove the db.

cpuguy83 commented:
All green

tiborvass commented on daemon/daemon.go:
```diff
@@ -332,63 +326,65 @@ func (daemon *Daemon) restore() error {
+	// migrate any legacy links from sqlite
+	linkdbFile := filepath.Join(daemon.root, "linkgraph.db")
+	var legacyLinkDB *graphdb.Database
+	if _, err = os.Stat(linkdbFile); err == nil {
```
@cpuguy83 couldn't we do the `container.HostConfig == nil || container.HostConfig.Links != nil` check in the `containers` loop above to detect if legacyLinkDB should be set at all?

cpuguy83 replied:
Yep!

tiborvass commented on daemon/daemon.go:
```diff
@@ -332,63 +326,65 @@ func (daemon *Daemon) restore() error {
-				continue
+		if legacyLinkDB != nil {
+			if err := daemon.migrateLegacySqliteLinks(legacyLinkDB, c); err != nil {
+				legacyLinkDB.Close()
```
no need for this, we already deferred it above.

tiborvass commented:
LGTM

LK4D4 commented:
LGTM
let's test this in real world :)

cpuguy83 commented:
ping @ibuildthecloud aka, demolisher of code, hopes, and dreams.

thaJeztah commented:
Yay, merged! Thanks for being so patient @cpuguy83 :metal:

vincentwoo commented:
Is this the candidate fix to #17720? If this is on experimental I'm down to see if it'll still lockup on my production boxes.

cpuguy83 commented:
@vincentwoo This definitely speeds up some pieces of `docker ps`, and container creation. I'm not sure that it would fix speed reductions over time.

vincentwoo commented:
I can give it a run. Is it on https://experimental.docker.com? I have a hard time figuring out what revision that thing is at.

cpuguy83 commented:
This is in stable.

vincentwoo commented:
Right, but there's not binaries out against it yet, right? Slated for 1.10?

cpuguy83 commented:
https://master.dockerproject.org

vincentwoo commented:
Niiiice.
